src/main.rs

```diff
@@ -435,6 +451,52 @@ impl HabitCtl {
         false
     }
 
+        fn habit_skipified(&self, habit: &Habit, date: &Date<Local>) -> bool {
+        if habit.every_days < 1 {
+            return false;
+        }
+
+        let from = date
+            .checked_sub_signed(chrono::Duration::days(habit.every_days - 1))
+            .unwrap();
+        let mut current = from;
+        while current <= *date {
+            if let Some(entry) = self.get_entry(&current, &habit.name) {
+                if entry.value == "s" {
+                    return true;
+                }
+            }
+            current = current
+                .checked_add_signed(chrono::Duration::days(1))
+                .unwrap();
+        }
+        false
+    }
+
+        fn habit_warning(&self, habit: &Habit, date: &Date<Local>) -> bool {
+        if habit.every_days < 1 {
+            return false;
+        }
+
+        let from = date
+            .checked_sub_signed(chrono::Duration::days(habit.every_days))
+            .unwrap();
+        let mut current = *date;
+        while current >= from {
+            if let Some(entry) = self.get_entry(&current, &habit.name) {
+                if (entry.value == "y" || entry.value == "s") && current - from > chrono::Duration::days(0) {
+                        return false;
+                } else if (entry.value == "y" || entry.value == "s") && current - from == chrono::Duration::days(0) {
+                    return true;
```

Comment:
So this starts warning when you're on your last day, right? Seems reasonable.

Reply:
Yup, this iterates backwrds from the current date which would normally only be a ? in the interface and sees if you are on the last day you can be before breaking the chain. If it is the last day it returns a true for Day::Warning and that is given a sigil of "!" in the interface. 
(it took me a whiel to figure this one out since I am not familiar with the Chrono library and Rust's type system. I had never programmed any Rust till I started hacking on this code base.  =p  )